genrl/deep/agents/dqn/categorical.py

```diff
@@ -0,0 +1,144 @@
+        layers (:obj:`tuple` of :obj:`int`): Layers in the Neural Network
+            of the Q-value function
+        lr_value (float): Learning rate for the Q-value function
+        replay_size (int): Capacity of the Replay Buffer
+        buffer_type (str): Choose the type of Buffer: ["push", "prioritized"]
+        max_epsilon (str): Maximum epsilon for exploration
+        min_epsilon (str): Minimum epsilon for exploration
+        epsilon_decay (str): Rate of decay of epsilon (in order to decrease
+            exploration with time)
+        noisy_layers (:obj:`tuple` of :obj:`int`): Noisy layers in the Neural
+            Network of the Q-value function
+        num_atoms (int): Number of atoms used in the discrete distribution
+        v_min (int): Lower bound of value distribution
+        v_max (int): Upper bound of value distribution
+        seed (int): Seed for randomness
+        render (bool): Should the env be rendered during training?
+        device (str): Hardware being used for training. Options:
+            ["cuda" -> GPU, "cpu" -> CPU]
+    """
+
+    def __init__(
+        self,
+        *args,
+        noisy_layers: Tuple = (32, 128),
+        num_atoms: int = 51,
+        v_min: int = -10,
+        v_max: int = 10,
+        **kwargs
+    ):
+        super(CategoricalDQN, self).__init__(*args, **kwargs)
+        self.noisy_layers = noisy_layers
+        self.num_atoms = num_atoms
+        self.v_min = v_min
+        self.v_max = v_max
+
+        self.empty_logs()
+        self.create_model()
+
+    def create_model(self, *args) -> None:
+        """Function to initialize Q-value model
+
+        Initialises Q-value mode based on network type ["cnn", "mlp"]
+        """
+        input_dim, action_dim, _, _ = get_env_properties(self.env, self.network_type)
+
+        self.model = get_model("dv", self.network_type + "categorical")(
+            input_dim, action_dim, self.layers, self.noisy_layers, self.num_atoms
+        )
+        self.target_model = deepcopy(self.model)
+
+        self.replay_buffer = self.buffer_class(self.replay_size, *args)
+        self.optimizer = opt.Adam(self.model.parameters(), lr=self.lr_value)
+
+    def select_action(
+        self, state: np.ndarray, deterministic: bool = False
+    ) -> np.ndarray:
+        """Select action given state
+
+        Epsilon-greedy action-selection
+
+        Args:
+            state (:obj:`np.ndarray`): Current state of the environment
+            deterministic (bool): Should the policy be deterministic or stochastic
+
+        Returns:
+            action (:obj:`np.ndarray`): Action taken by the agent
+        """
+        if not deterministic:
+            if np.random.rand() < self.epsilon:
+                return np.asarray(self.env.sample())
+
+        state = torch.FloatTensor(state)
+        dist = self.model(state).data.cpu()
+        dist = dist * torch.linspace(self.v_min, self.v_max, self.num_atoms)
+        action = dist.sum(2).max(1)[1].numpy()
```

Comment:
Can you add a comment explaining this operation? Makes it more readable as there a bunch of operations. And it would help in debugging

Reply:
Yeah ok